File: src/event.ts
import { components } from '@octokit/openapi-types';
import * as fs from 'fs';
import { log } from './logger';

export type Event = {
  before: string | null;
  after: string | null;
  pull_request: components['schemas']['pull-request'] | null;
  app: components['schemas']['nullable-integration'];
  repository: components['schemas']['minimal-repository'];
  number?: number;
};

const readEvent = (): Event | undefined => {
  try {
    if (process.env.GITHUB_EVENT_PATH) {
      return JSON.parse(fs.readFileSync(process.env.GITHUB_EVENT_PATH, 'utf8'));
    }
  } catch (e) {}
};

export const getEvent = (): Event => {
  const event = readEvent();

  log.debug(`event = `, event);

  if (!event) {
    throw new Error('Failed to get github event.json.');
  }

  return event;
};

File: src/event.rs
use anyhow::{anyhow, Context, Result};
use serde::Deserialize;

#[derive(Debug, Clone, Deserialize)]
pub struct Event {
    pub before: Option<String>,
    pub after: Option<String>,
    pub number: Option<u64>,
    #[serde(default)]
    pub pull_request: Option<PullRequest>,
    #[serde(default)]
    pub repository: Option<MinimalRepository>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct PullRequest {
    pub number: u64,
    pub head: GitRef,
    pub base: GitRef,
}

#[derive(Debug, Clone, Deserialize)]
pub struct GitRef {
    pub sha: String,
    #[serde(rename = "ref")]
    pub ref_name: Option<String>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct MinimalRepository {
    pub name: Option<String>,
    pub full_name: Option<String>,
    pub default_branch: Option<String>,
}

pub fn get_event() -> Result<Event> {
    let path = std::env::var("GITHUB_EVENT_PATH")
        .map_err(|_| anyhow!("Failed to get github event.json. GITHUB_EVENT_PATH is not set."))?;
    let raw = std::fs::read_to_string(&path)
        .with_context(|| format!("Failed to read event.json at {path}"))?;
    let event: Event = serde_json::from_str(&raw)
        .with_context(|| format!("Failed to parse event.json at {path}"))?;
    tracing::debug!(?event, "github event");
    Ok(event)
}
